syncapi/internal/history_visibility.go

```diff
@@ -0,0 +1,204 @@
+	})
+	if len(events) == 0 {
+		return events, nil
+	}
+	start := time.Now()
+
+	// try to get the current membership of the user
+	membershipCurrent, _, err := syncDB.SelectMembershipForUser(ctx, events[0].RoomID(), userID, math.MaxInt64)
+	if err != nil {
+		return nil, err
+	}
+
+	eventIDs := make([]string, len(events))
+	for i := range events {
+		eventIDs[i] = events[i].EventID()
+	}
+
+	// Get the mapping from eventID -> eventVisibility
+	eventsFiltered := make([]*gomatrixserverlib.HeaderedEvent, 0, len(events))
+	event, err := visibilityForEvents(ctx, rsAPI, eventIDs, userID, events[0].RoomID())
+	if err != nil {
+		return eventsFiltered, err
+	}
+	for _, ev := range events {
+		d := event[ev.EventID()]
+		d.membershipCurrent = membershipCurrent
+		d.visibility = ev.Visibility
+		// Always include specific state events for /sync responses
+		if alwaysIncludeEventIDs != nil {
+			if _, ok := alwaysIncludeEventIDs[ev.EventID()]; ok {
+				eventsFiltered = append(eventsFiltered, ev)
+				continue
+			}
+		}
+		// NOTSPEC: Always allow user to see their own membership events (spec contains more "rules")
+		if ev.Type() == gomatrixserverlib.MRoomMember && ev.StateKey() != nil && *ev.StateKey() == userID {
+			eventsFiltered = append(eventsFiltered, ev)
+			continue
+		}
+		// Handle history visibility changes
+		if hisVis, err := ev.HistoryVisibility(); err == nil {
+			prevHisVis := gjson.GetBytes(ev.Unsigned(), "prev_content.history_visibility").String()
+			if oldPrio, ok := historyVisibilityPriority[gomatrixserverlib.HistoryVisibility(prevHisVis)]; ok {
+				// no OK check, since this should have been validated when setting the value
+				newPrio := historyVisibilityPriority[hisVis]
+				if oldPrio < newPrio {
+					d.visibility = gomatrixserverlib.HistoryVisibility(prevHisVis)
+				}
+			}
+		}
+		// do the actual check
+		allowed := d.allowed()
+		if allowed {
+			eventsFiltered = append(eventsFiltered, ev)
+		}
+	}
+	calculateHistoryVisibilityDuration.With(prometheus.Labels{"api": endpoint}).Observe(float64(time.Since(start).Milliseconds()))
+	return eventsFiltered, nil
+}
+
+// visibilityForEvents returns a map from eventID to eventVisibility containing the visibility and the membership at the given event.
+// Returns an error if the roomserver can't calculate the memberships.
+func visibilityForEvents(ctx context.Context, rsAPI api.SyncRoomserverAPI, eventIDs []string, userID, roomID string) (map[string]eventVisibility, error) {
+	res := make(map[string]eventVisibility, len(eventIDs))
+
+	// get the membership events for all eventIDs
+	resp := &api.QueryMembersipAtEventResponse{}
+	err := rsAPI.QueryMembershipAtEvent(ctx, &api.QueryMembersipAtEventRequest{
+		RoomID:   roomID,
+		EventIDs: eventIDs,
+		UserID:   userID,
+	}, resp)
+	if err != nil {
+		return res, err
+	}
+
+	// Create a map from eventID -> eventVisibility
+	for _, eventID := range eventIDs {
+		vis := eventVisibility{membershipAtEvent: gomatrixserverlib.Leave}
```

Comment:
Updated.
Also, it should never be 0 (besides if we manually update the visibility of an event), since the `sql.Scanner`/`sql.Valuer` from https://github.com/matrix-org/gomatrixserverlib/pull/326/files default to `shared`.